feat: handle cancel messages

vasco-santos opened the pull request:
This PR adds handling of cancel messages from inflight operations. Using `p-queue` together with Abort Signal all the way would be my ideal solution, but that would require a huge refactor that would mean a long time to work on this.

Implementation details:
- added `inProcessingWantBlocks` and `inProcessingWantHaves` Maps to the handler context (each connection has its own context)
- decoupled previous `batchFetch` function into two function, with an initial step just for getting a normalized wantList with arrays for `wantedBlocks`, `wantedHave` and `canceled`
- `wantedBlocks` and `wantedHave` added to respective processing maps in the start of fetching, and canceled blocks flagged in respective processing maps (if still existing and not previously addressed)
- cleanup `inProcessing` maps in case of errors and messages not handled to avoid memory leaks

Closes https://github.com/web3-storage/w3infra/issues/149

vasco-santos commented on test/handler.test.js:
```diff
@@ -194,14 +194,38 @@ t.test('handle', async t => {
     await handle({ context: contextSpy, logger: loggerSpy, batchSize: 1 })
 
-    t.equal(connectionSpy.send.callCount, 2) // Only two sends in the 4 block messages
+    t.equal(connectionSpy.send.callCount, 0) // Blocks requested and cancelled in same message
```
The mock actually intercepts the HTTP request from client, and given async operation means now we actually cancel everything 🎉

vasco-santos commented on test/handler.test.js:
```diff
@@ -307,18 +331,14 @@ t.test('handle', async t => {
       cid: 'bafkreihtgll4nt4euynqnkjrg3vclmmmvo7srby4qul6pc4uiiayorplmu',
       data: 'cid8-content'
     }))
-    t.ok(responseContainsData(response, {
-      cid: 'bafkreia6rl74tk7lwetxgol4pjbknva4474ecxkx7vlhviz7xpsfascqhu',
-      data: 'cid1-content'
-    }))
```
third batch has wantBlock and wantBlock cancel

alanshaw reviewed (changes requested):
I don't understand how this works. We create a new `context` for every bitswap message we recieve. So additional messages that are received on the same connection won't affect any in-flight responses...

vasco-santos commented:
> I don't understand how this works. We create a new context for every bitswap message we recieve. So additional messages that are received on the same connection won't affect any in-flight responses...

Yes 🤦🏼  I moved the Maps creation into `handler.js` to not initialise them in the tests too and did not notice it was being created in message scope. I would expect the context to be connection dependent but will keep as is and just moved tracking of in flight messages to be in connection scope, but out of message scope.

alanshaw commented on src/handler.js:
```diff
@@ -15,55 +15,66 @@ function createContext ({ service, peerId, protocol, wantlist, awsClient, connec
             }
           } catch (err) {
+            // Cleanup in processing blocks
+            clearProcessingBlocks(normalizedWantlist, context)
```
I'm concerned we'll end up with a memory leak for canceled blocks that we already sent. Perhaps it should be an LRU also?

alanshaw reviewed (changes requested):
🚀 this is awesome, there's just a few issues I think we need to iron out before merge.

vasco-santos commented:
@alanshaw thanks for the review. Integrated your suggestions:
- single LRU per connection (value has the cancel type because there are no )
  - still not entirely sure. In practise, we should not receive at same time cancel for block and want and this should work fine. But if not the case, there will happen overwrites... To solve this, we need to go back to 2 data structures or play with key format. What do you think?
- added telemetry for count of success cancels and size. Followed pattern for metrics as commented `bitswap-request-size` here
- clear on disconnect/error

alanshaw reviewed (approved):
LGTM, please can we just use the existing LRU module though?

vasco-santos commented on src/service.js:
```diff
@@ -4,6 +4,7 @@ import { webSockets } from '@libp2p/websockets'
 import { noise } from '@chainsafe/libp2p-noise'
 import { mplex } from '@libp2p/mplex'
 // import { yamux } from '@chainsafe/libp2p-yamux'
+import LRU from 'quick-lru'
```
good catch

vasco-santos commented:
Deployed to dev https://github.com/elastic-ipfs/bitswap-peer/actions/runs/4305195468 and ran regression tests https://github.com/elastic-ipfs/bitswap-e2e-tests/actions/runs/4305374528 successfully 🎉
